feat(Maintenance Schedule): send_maintenance_schedule_reminder_notifi…

Murtazaabidi1 opened the pull request:
…cation

Murtazaabidi1 committed:
feat(Maintenance Schedule): send_maintenance_schedule_reminder_notification

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -66,6 +67,45 @@ def adjust_scheduled_date_for_holiday(self, scheduled_date):
 
+	def send_maintenance_schedule_reminder_notification(self, row_name):
+		msd_doctype = "Maintenance Schedule Detail"
+		row = [d for d in self.schedules if d.name == row_name][0]
```
Not safe... Handle row not found case

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -66,6 +67,45 @@ def adjust_scheduled_date_for_holiday(self, scheduled_date):
+				return False
+
+		# Maintenance Schedule scheduled Date is in the past
+		if notification_type in ("Maintenance Reminder"):
```
Commonify if block

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -66,6 +67,45 @@ def adjust_scheduled_date_for_holiday(self, scheduled_date):
+
+			if maintenance_schedule_row.scheduled_date < getdate():
+				if throw:
+					frappe.throw(_("Cannot send {0} notification after Scheduled Time has passed")
```
after Scheduled **Date** has passed *

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -66,6 +67,45 @@ def adjust_scheduled_date_for_holiday(self, scheduled_date):
+		if notification_type in ("Maintenance Reminder"):
+			maintenance_schedule_row = [d for d in self.schedules if d.name == child_name][0]
+
+			if maintenance_schedule_row.scheduled_date < getdate():
```
wrap `getdate(maintenance_schedule_row.scheduled_date)` to be sure that scheduled_date also check for null

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -66,6 +67,45 @@ def adjust_scheduled_date_for_holiday(self, scheduled_date):
+
+		# Maintenance Schedule scheduled Date is in the past
+		if notification_type in ("Maintenance Reminder"):
+			maintenance_schedule_row = [d for d in self.schedules if d.name == child_name][0]
```
Not safe .. handle row not found case

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +317,34 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+
+
+@frappe.whitelist()
+def send_maintenance_schedule_reminder_notification():
```
`send_maintenance_schedule_reminder_notifications` plural

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +317,34 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+	schedule_to_remind = frappe.db.sql("""
+		SELECT ms.name AS ms_name, msd.name AS row_name, msd.scheduled_date
+		FROM `tabMaintenance Schedule` ms
+		LEFT JOIN `tabMaintenance Schedule Detail` msd ON msd.parent = ms.name
```
use `INNER JOIN`

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +317,34 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+			And nc.child_doctype = 'Maintenance Schedule Detail' AND nc.child_name = msd.name
+		WHERE ms.status = "Active"
+			AND msd.scheduled_date <= %(target_date)s
+			AND nc.last_scheduled_dt is null
```
last sent date condition?

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +317,34 @@ def create_maintenance_opportunity(maintenance_schedule, row):
 	return target_doc
+
+
+@frappe.whitelist()
```
Remove whitelist when done with testing

Murtazaabidi1 committed:
chore: requested PR changes

Murtazaabidi1 committed:
chore: requested PR changes
- Maintenance Reminder Notification Time of Day (Added)
- Split functions
- Query review
- Automated reminder enabled condition

Murtazaabidi1 committed:
Merge branch 'version-14' of https://github.com/ParaLogicTech/erpnext into version-14

# Conflicts:
#	erpnext/patches.txt

SaiFi0102 commented on erpnext/hooks.py:
```diff
@@ -277,6 +277,7 @@
 		"erpnext.healthcare.doctype.patient_appointment.patient_appointment.set_appointment_reminder",
 		"erpnext.vehicles.doctype.vehicle_booking_order.vehicle_booking_order.send_payment_overdue_notifications",
 		"erpnext.crm.doctype.appointment.appointment.send_appointment_reminder_notifications",
+        "erpnext.crm.doctype.maintenance_schedule.maintenance_schedule.send_maintenance_schedule_reminder_notifications",
```
Use tabs, not spaces

SaiFi0102 commented on erpnext/crm/doctype/crm_settings/crm_settings.json:
```diff
@@ -14,7 +15,11 @@
-  "auto_mark_opportunity_as_lost"
+  "auto_mark_opportunity_as_lost",
+  "column_break_q3twi",
+  "maintenance_reminder_days_before_schedule",
```
Lets keep fieldnames consistent with what we already have.
Rename fieldname to `maintenance_reminder_days_before`

SaiFi0102 commented on erpnext/crm/doctype/crm_settings/crm_settings.json:
```diff
@@ -14,7 +15,11 @@
+  "column_break_q3twi",
+  "maintenance_reminder_days_before_schedule",
+  "maintenance_reminder_time",
+  "auto_reminder_for_maintenance_schedule"
```
Lets remove this checkbox for now, since this is already being checked with SMS Template

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +327,70 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+	if notification_last_sent_date and getdate(notification_last_sent_date) >= reminder_date:
+		return
+
+	remind_before_days = cint(frappe.db.get_single_value("CRM Settings", "maintenance_reminder_days_before_schedule"))
```
rename to `remind_days_before`

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +327,70 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+
+	for d in schedule_to_remind:
+		doc = frappe.get_doc("Maintenance Schedule", d.ms_name)
+		doc.send_maintenance_schedule_reminder_notification(d.msd_name)
```
it should be `d.row_name` see query below

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +327,70 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+
+	frappe.db.set_default("maintenance_schedule_notification_last_sent_date", reminder_date)
+
+def get_schedules_to_remind(target_date):
```
rename to `get_maintenance_schedules_for_reminder_notification`

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +327,70 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+	}, as_dict=1)
+	return schedule_to_remind
+
+def automated_reminder_enabled():
```
rename to `automated_maintenance_reminder_enabled`

SaiFi0102 commented on erpnext/patches/v14_0/update_project_template_name_in_maintenance_schedule_detail.py:
```diff
@@ -0,0 +1,7 @@
+def execute():
+	frappe.reload_doc("maintenance", "doctype", "maintenance_schedule_detail")
+
+	frappe.db.sql("""update `tabMaintenance Schedule Detail` as msd  set msd.project_template_name = \
```
Better we use inner join

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +327,70 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+		WHERE ms.status = 'Active'
+			AND msd.scheduled_date <= %(target_date)s
+			AND nc.last_scheduled_dt is NULL
+			AND nc.last_sent_dt is NULL
```
Change condition like used in `get_appointments_for_reminder_notification`

Murtazaabidi1 committed:
chore: requested PR changes

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -5,9 +5,10 @@
 from erpnext.utilities.transaction_base import TransactionBase
 from dateutil.relativedelta import relativedelta
-from frappe.utils import add_days, getdate, cstr, today
+from frappe.utils import add_days, getdate,get_time, now_datetime,combine_datetime, cstr, today, cint
```
space after comma

SaiFi0102 commented on erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:
```diff
@@ -277,3 +327,66 @@ def create_maintenance_opportunity(maintenance_schedule, row):
+		WHERE ms.status = 'Active'
+			AND msd.scheduled_date = %(target_date)s
+			AND nc.last_scheduled_dt is NULL
+			AND (nc.last_sent_dt is null or DATE(nc.last_sent_dt) != %(target_date)s)
```
add condition for `schedule_date >= today date`

Murtazaabidi1 committed:
chore: requested PR changes

Murtazaabidi1 committed:
Merge branch 'version-14' of https://github.com/ParaLogicTech/erpnext into version-14

# Conflicts:
#	erpnext/hooks.py
#	erpnext/patches.txt
